modules/siddhi-core/src/main/java/io/siddhi/core/table/CacheTableLFU.java

```diff
@@ -0,0 +1,234 @@
+            readWriteLock.writeLock().unlock();
+        }
+    }
+
+    @Override
+    public void updateOrAddWithMaxSize(ComplexEventChunk<StateEvent> updateOrAddingEventChunk,
+                                       CompiledCondition compiledCondition,
+                                       CompiledUpdateSet compiledUpdateSet,
+                                       AddingStreamEventExtractor addingStreamEventExtractor, int maxTableSize) {
+        ComplexEventChunk<StateEvent> updateOrAddingEventChunkForCache = new ComplexEventChunk<>(true);
+        updateOrAddingEventChunk.reset();
+        while (updateOrAddingEventChunk.hasNext()) {
+            StateEvent event = updateOrAddingEventChunk.next();
+            updateOrAddingEventChunkForCache.add((StateEvent) generateEventWithRequiredFields(event, siddhiAppContext,
+                    cacheExpiryEnabled));
+        }
+        readWriteLock.writeLock().lock();
+        TableState state = stateHolder.getState();
+        try {
+            String primaryKey;
+            if (stateHolder.getState().getEventHolder() instanceof IndexEventHolder) {
+                updateOrAddingEventChunkForCache.reset();
+                while (updateOrAddingEventChunkForCache.hasNext()) {
+                    StateEvent matchingEvent = updateOrAddingEventChunkForCache.next();
+                    IndexEventHolder indexEventHolder = (IndexEventHolder) stateHolder.getState().getEventHolder();
+                    primaryKey = getPrimaryKey(compiledCondition, matchingEvent);
+                    StreamEvent usedEvent = indexEventHolder.getEvent(primaryKey);
+                    if (usedEvent != null) {
+                        int newUsage = (int) usedEvent.getOutputData()[usedEvent.getOutputData().length - 1] + 1;
+//                        usedEvent.getOutputData()[usedEvent.getOutputData().length - 1] = newUsage;
+                        matchingEvent.getStreamEvent(0).getOutputData()[matchingEvent.getStreamEvent(0).
+                                getOutputData().length - 1] = newUsage;
+                    }
+                }
+            }
+            ComplexEventChunk<StreamEvent> failedEvents = ((Operator) compiledCondition).tryUpdate(
+                    updateOrAddingEventChunkForCache,
+                    state.getEventHolder(),
+                    (InMemoryCompiledUpdateSet) compiledUpdateSet,
+                    addingStreamEventExtractor);
+            if (failedEvents != null && failedEvents.getFirst() != null) {
+                int tableSize = this.size();
+                ComplexEventChunk<StreamEvent> failedEventsLimitCopy = new ComplexEventChunk<>();
+                failedEvents.reset();
+                while (true) {
+                    failedEventsLimitCopy.add(failedEvents.next());
+                    tableSize++;
+                    if (tableSize == maxTableSize || !failedEvents.hasNext()) {
+                        break;
+                    }
+                }
```

Comment:
optimize line 163 to 171 
also adding without copying/or removing from the check will cause issues please fix this in all places.